rocDecode: Added decode buffer pool implementation for AVC and HEVC.

jeffqjiangNew opened the pull request:
- Now DPB buffer management is decoupled from decode buffer pool management.
- User can now specify the decode buffer pool size and set display delay parameter.
- Conformance tests on MI250/MI300/Navi31 remain passing.
- Tests on MI250/MI300A/Navi31 showed no performance regressions.
- Multiple session and very long session stability tests on MI250/MI300A/Navi31 showed no issues.

jeffqjiangNew committed:
* rocDecode: Initial check in for decode buffer pool.

jeffqjiangNew committed:
* rocDecode: All 135 streams pass.

jeffqjiangNew committed:
* rocDecode: Fixed a build error in debug mode.

jeffqjiangNew committed:
* rocDecode/HEVC: Removed two workaround in HEVC DPB management, after decode buffer pool implementa
tion.
 - WR 1: Conditional bumping (when max_num_reorder_pics > 0) to avoid synchronous job submission in
C.5.2.3.
 - WR 2: Add two more buffers in DPB to avoid buffer over-writing.

jeffqjiangNew committed:
* rocDecode/HEVC: Added display delay feature.

jeffqjiangNew committed:
* rocDecode/HEVC: Fixed the -z option issue within the context of the new decode buffer pool implementation.

jeffqjiangNew committed:
* rocDecode/HEVC: Removed redundent code.

jeffqjiangNew committed:
* rocDecode/AVC: Added decode buffer pool implementation for AVC.

jeffqjiangNew committed:
* rocDecode: Added a few changes.
 - Added display delay feature to AVC.
 - Removed a workaround for AVC: AVC_MAX_DPB_FRAMES was increased to 18. Now it is back to 16.
 - Removed a workaround for AVC: increased DPB buffer size by 2. Now it is back to normal.
 - Code format changes for HEVC.

jeffqjiangNew committed:
* rocDecode/AVC: Fixed the -z option issue within the context of the new decode buffer pool implementation.

jeffqjiangNew committed:
* rocDecode: Merged OutputDecodedPictures() method to upper class RocVideoParser.

jeffqjiangNew committed:
* rocDecode: Code cleanup. No functional changes.

AryanSalmanpour commented on utils/rocvideodecode/roc_video_dec.cpp:
```diff
@@ -39,9 +39,10 @@ RocVideoDecoder::RocVideoDecoder(int device_id, OutputSurfaceMemoryType out_mem_
     parser_params.codec_type = codec_id_;
-    parser_params.max_num_decode_surfaces = 1;
+    // Note this is tunable
+    parser_params.max_num_decode_surfaces = MAX_DPB_FRAMES + DISPLAY_DELAY;
```
The current change adjusts the buffer pool to 18, resulting in additional unused memory being allocated per stream. Prior to this pull request, 7 surfaces were allocated to decode the AMD_driving_virtual_20-H265.mp4 sample. However, after the pull request, we are now allocating 18 surfaces, which represents an additional overhead.



**The current ToT develop branch:**
./videodecode -i ../../../data/videos/AMD_driving_virtual_20-H265.mp4 
info: Input file: AMD_driving_virtual_20-H265.mp4
info: Using GPU device 0 - AMD Radeon RX 6750 XT[gfx1031] on PCI bus 0b:00.0
info: decoding started, please wait!
Input Video Information
        Codec        : H.265/HEVC
        Frame rate   : 30000/1001 = 29.97 fps
        Sequence     : Progressive
        Coded size   : [2048, 1024]
        Display area : [0, 0, 2048, 1024]
        Chroma       : YUV 420
        Bit depth    : 8
Video Decoding Params:
        **Num Surfaces : 7**
        Crop         : [0, 0, 2048, 1024]
        Resize       : 2048x1024

info: Total frame decoded: 601
info: avg decoding time per frame: 1.79221 ms
info: avg FPS: 557.972

**The current PR:**
./videodecode -i ../../../data/videos/AMD_driving_virtual_20-H265.mp4 
info: Input file: AMD_driving_virtual_20-H265.mp4
info: Using GPU device 0 - AMD Radeon RX 6750 XT[gfx1031] on PCI bus 0b:00.0
info: decoding started, please wait!
Input Video Information
        Codec        : H.265/HEVC
        Frame rate   : 30000/1001 = 29.97 fps
        Sequence     : Progressive
        Coded size   : [2048, 1024]
        Display area : [0, 0, 2048, 1024]
        Chroma       : YUV 420
        Bit depth    : 8
Video Decoding Params:
        **Num Surfaces : 18**
        Crop         : [0, 0, 2048, 1024]
        Resize       : 2048x1024

info: Total frame decoded: 601
info: avg decoding time per frame: 1.61563 ms
info: avg FPS: 618.955

jeffqjiangNew replied:
Right. The size 18 of the decode buffer pool is set by the user. In our case, the user is the util layer. It can be tuned. I have a suggestion that we can discuss in the review meeting.

AryanSalmanpour replied:
Why should the user decide on the decode buffer pool? It should be determined by the parser.

jeffqjiangNew replied:
Because the user can determine how much buffering it wants before display. Note decode buffer pool is not the same as DPB.

jeffqjiangNew replied:
We probably want to use the display delay as the tuning parameter. Which is my suggestion.

jeffqjiangNew replied:
Briefly, the suggestion is to set the decode buffer pool size to DPB size + display delay. The current hard coded 18 only considers the worst scenario.

jeffqjiangNew replied:
Commit 2ad9e01 should address this issue. I'll push another commit for other small changes.

jeffqjiangNew committed:
* rocDecode: Made decode buffer pool size adaptive.
 - Removed the hard coded decode buffer pool size set in the decoder.
 - Exposed the display delay parameter from RocVideoDecoder class to the user.
 - Now the decoder buffer pool size is determined from the DPB buffer size and display delay parameter.

jeffqjiangNew committed:
* rocDecode: Several changes based on code review.
 - Merged decode and display use status into one parameter.
 - Removed the surface index from DecodeFrameBuffer, which is now implicitly referred by the array index.
 - Changed a function name for better clarity.

jeffqjiangNew committed:
* rocDecode: Added a comment.

rrawther commented on src/parser/roc_video_parser.cpp:
```diff
@@ -57,16 +57,65 @@ rocDecStatus RocVideoParser::Initialize(RocdecParserParams *pParams) {
+    num_output_pics_ = 0;
+}
+
+void RocVideoParser::CheckAndAdjustDecBufPoolSize(int dpb_size) {
```
where is this function called

rrawther commented on src/parser/roc_video_parser.cpp:
```diff
@@ -57,16 +57,65 @@ rocDecStatus RocVideoParser::Initialize(RocdecParserParams *pParams) {
+    }
+}
+
+ParserResult RocVideoParser::OutputDecodedPictures(bool no_delay) {
```
where will this be called?
